numpyro/distributions/directional.py

```diff
@@ -1,22 +1,79 @@
 # Copyright Contributors to the Pyro project.
 # SPDX-License-Identifier: Apache-2.0
 
+from collections import namedtuple
+import functools
 import math
+from math import pi
+import operator
 
 from jax import lax
 import jax.numpy as jnp
 import jax.random as random
-from jax.scipy.special import erf, i0e, i1e
+from jax.scipy.special import erf, i0e, i1e, logsumexp
 
 from numpyro.distributions import constraints
 from numpyro.distributions.distribution import Distribution
 from numpyro.distributions.util import (
     is_prng_key,
+    lazy_property,
     promote_shapes,
     safe_normalize,
     validate_sample,
     von_mises_centered,
 )
+from numpyro.util import while_loop
+
+
+def _numel(shape):
+    return functools.reduce(operator.mul, shape, 1)
+
+
+def log_I1(orders: int, value, terms=250):
+    r"""Compute first n log modified bessel function of first kind
+    .. math ::
+        \log(I_v(z)) = v*\log(z/2) + \log(\sum_{k=0}^\inf \exp\left[2*k*\log(z/2) - \sum_kk^k log(kk)
+        - \lgamma(v + k + 1)\right])
+    :param orders: orders of the log modified bessel function.
+    :param value: values to compute modified bessel function for
+    :param terms: truncation of summation
+    :return: 0 to orders modified bessel function
+    """
+    orders = orders + 1
+    if value.ndim == 0:
+        vshape = jnp.shape([1])
+    else:
+        vshape = value.shape
+    value = value.reshape(-1, 1)
+    flat_vshape = _numel(vshape)
+
+    k = jnp.arange(terms)
+    lgammas_all = lax.lgamma(jnp.arange(1.0, terms + orders + 1))
+    assert lgammas_all.shape == (orders + terms,)  # lgamma(0) = inf => start from 1
+
+    lvalues = lax.log(value / 2) * k.reshape(1, -1)
```

Comment:
Use jnp.log. Basically, we will try to avoid lax api as much as possible.

Reply:
I see; I'll remove `lax` throughout.